Tips to upload large models/datasets

Wauplin opened the pull request:
We've talked several times about having a section in the docs with some tips for users wanting to upload a large amount of data. I tried to sum-up everything with two distinct aspects:
- technical limitations (max size per file, max file per commit, max file per folder, max file per repo)
- practical tips (start small, use hf_transfer, expect failures)

Wauplin commented:
cc @Pierrci @coyotte508 can you check the info is accurate. I took it from https://github.com/huggingface/huggingface_hub/issues/995#issuecomment-1218396702 but prefer to double-check it's not outdated (already 1y old :smile:)

cc @lhoestq might be interested for datasets users as well

lhoestq commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+(e.g. an image dataset). Each repository will have its own limitations so here is a list of factors to consider:
+
+- **Repository size**: The total size of the data you're planning to upload. There is no hard limit on the size of a repository on
+Hub. However if you plan to upload hundreds of GBs or even TBs of data, we would be grateful if you could let us know
```
"let us know in advance" -> how ? maybe suggest the discord server for example

osanseviero replied:
Or email

Wauplin replied:
@osanseviero which email address would you suggest? Maybe datasets@huggingface.co ?

julien-c replied:
yes datasets sounds good to me (website@ also an option)

lhoestq reviewed (approved):
Awesome !

stevhliu commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+      happen when streaming data and smaller files avoid resuming from the beginning in case of errors.
+    - Files are served to the users using CloudFront. From our experience, files above 30GB are not cached by this service
+      leading to a slower download speed.
+    - A hard-limit of 50GB has been set on the Hub. If you try to commit larger files, an error will be returned by the server.
```
```suggestion
    - A hard-limit of 50GB has been set on the Hub for an individual file. If you try to commit larger files, an error will be returned by the server.
```

stevhliu commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+fresh start.
+- **Number of operations per commit**: Once again, there is no hard-limit here. When a commit is done on the Hub, each
+git operation (addition or delete) is checked by the server before actually doing it. This means that when a hundred LFS
+files are committed at once, the server must check if each file has been correctly uploaded -checking its existence
+and size-. A timeout of 60s is set on the request, meaning that if the process takes more time an error is raised
```
```suggestion
files are committed at once, the server must check if each file has been correctly uploaded, checking its existence
and size. A timeout of 60s is set on the request, meaning that if the process takes more time an error is raised
```

stevhliu commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+to add around 50-100 files per commit.
+
+To summarize it quickly:
+- Reach out to us for large repos (TBs of data)
```
I think you could make a table out of this and stick it at the beginning of the section right before "Technical limitations" so users can quickly scan it. It would also help to do something like:

| Limitation               | Recommendation              |
|--------------------------|-----------------------------|
| Large repo (TBs of data) | contact us                  |
| Max 1M files/repo        | merge data into fewer files |
| Max 10k files/folder     | use subdirectories in repo  |

osanseviero replied:
Great idea!

stevhliu reviewed (approved):
Awesome! Once merged, it may be nice to link to this from the Datasets docs as well 😄

Wauplin commented:
Thanks for the review and valuable feedback @stevhliu @lhoestq! I'll update to PR to make the short list more explicit about what is a hard limit and what is a recommendation. And also move it up in the section :)

osanseviero commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
 For more detailed information, take a look at the [`HfApi`] reference.
 
-## Push files with Git LFS
+## Tips and tricks for large uploads
```
Let's be careful to not break this url

Wauplin replied:
I would not be too worried about it to be honest. I've checked and `#upload-files-with-git-lfs` is referenced nowhere in our internal docs (both hfh and hf_docs). Doesn't mean that such a URL doesn't exist in the wild but I would expect probability to be quite low. And even if it's the case, users will be redirected to the correct page even though it's not the correct section. Since Git LFS upload is mostly deprecated _it should be fine_

osanseviero commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+
+### Technical limitations
+
+What are we talking about when we say "large uploads" and what are their associated limitations? Large uploads can be
```
```suggestion
What are we talking about when we say "large uploads", and what are their associated limitations? Large uploads can be
```

osanseviero commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+
+What are we talking about when we say "large uploads" and what are their associated limitations? Large uploads can be
+very diverse, from repositories with a few huge files (e.g. model weights) to repositories with thousands of small files
+(e.g. an image dataset). Each repository will have its own limitations so here is a list of factors to consider:
```
```suggestion
(e.g. an image dataset). Each repository will have its own limitations, so here is a list of factors to consider:
```

osanseviero commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +376,89 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+Now that we saw the technical aspects you must consider when structuring your repository, let's see some practical
+tips to make your upload process as smooth as possible.
+
+- **Start small**: We recommend to start with a small amount of data to test your upload script. It's easier to iterate
+on a script when failing takes only a small amount of time.
+- **Expect failures**: Streaming large amounts of data is challenging. You don't know what can happen but it's always
+best to consider that something will fail at least once -no matter if it's due to your machine, your connection or our
+servers. For example if you plan to upload a large number of files, it's best to keep track locally of which files you
+already uploaded before uploading the next batch. You are ensured that an LFS file that is already committed will never
+be re-uploaded twice but checking it client-side can still save some time.
+- **Use `hf_transfer`**: this is a Rust-based [library](https://github.com/huggingface/hf_transfer) meant to speed-up
+uploads on machines with very high bandwidth. To use it you must install it (`pip install hf_transfer`) and enable it
+by setting `HF_HUB_ENABLE_HF_TRANSFER=1` as environment variable. You can then use `huggingface_hub` normally.
+Disclaimer: this is a power user tool. It is tested and production-ready but lacks user-friendly features like progress
+bars or advanced error handling.
```
```suggestion
- **Start small**: We recommend starting with a small amount of data to test your upload script. It's easier to iterate
on a script when failing takes only a small amount of time.
- **Expect failures**: Streaming large amounts of data is challenging. You don't know what can happen, but it's always
best to consider that something will fail at least once -no matter if it's due to your machine, your connection, or our
servers. For example, if you plan to upload a large number of files, it's best to keep track locally of which files you
already uploaded before uploading the next batch. You are ensured that an LFS file that is already committed will never
be re-uploaded twice but checking it client-side can still save some time.
- **Use `hf_transfer`**: this is a Rust-based [library](https://github.com/huggingface/hf_transfer) meant to speed-up
uploads on machines with very high bandwidth. To use it, you must install it (`pip install hf_transfer`) and enable it
by setting `HF_HUB_ENABLE_HF_TRANSFER=1` as an environment variable. You can then use `huggingface_hub` normally.
Disclaimer: this is a power user tool. It is tested and production-ready but lacks user-friendly features like progress
bars or advanced error handling.
```

osanseviero reviewed (approved):
Cool stuff! :fire:

Wauplin commented:
I have addressed all the comments:
- merged the PR suggestions
- suggest to contact us via Discord or datasets@huggingface.co (I expect repos with TBs of data to be 99% datasets)
- compiled technical recommendation in a table as suggested by @stevhliu:

![image](https://github.com/huggingface/huggingface_hub/assets/11801849/8d542771-d328-43ef-86e4-c0901cf819a6)

@stevhliu Could you have a second look [to the docs](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1565/en/guides/upload#tips-and-tricks-for-large-uploads) and check it looks good? Thanks in advance!

julien-c commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +377,95 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+<Tip warning={true}>
+
+Although [`Repository`] is not formally deprecated, we recommend using the HTTP-based methods described above instead.
+For more details about this recommendation, please have a look to [this guide](../concepts/git_vs_http) explaining the
```
```suggestion
For more details about this recommendation, please have a look at [this guide](../concepts/git_vs_http) explaining the
```

julien-c reviewed (commented):
let's wait for @Pierrci quick review before merging please 🙏

stevhliu reviewed (approved):
Nice, love the additional columns you added to the table! 🤗

Wauplin commented on docs/source/guides/upload.md:
```diff
@@ -371,11 +377,95 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+completed server-side. This can be checked manually by browsing the repo on the Hub. To prevent this timeout, we recommend
+to add around 50-100 files per commit.
+
+To summarize it quickly:
```
Yes indeed!

Pierrci reviewed (commented):
I was finally able to take a look 😄

I don't think we should advertise hard limits, if you reach them it means your repo is going to be very hard to manage (both for the user and for us); we really want to steer people as much as we can toward numbers lower than those. We also want to keep some leeway and be able to change those limits at our discretion to protect our infra if needed.

So I pushed suggestions that advertise recommendations instead - and as for any recommendations, you can choose to ignore them if you're a player, but it's usually not a good idea...

Wauplin commented:
Thanks for the review @Pierrci! Completely understand your point here about not to commit to much on the hard limits. I have applied your recommendations. Could you have a final look and we are good to merge? :)

Pierrci commented:
Thanks @Wauplin! I did a Grammarly pass and pushed additional changes (the [VSCode extension](https://marketplace.visualstudio.com/items?itemName=znck.grammarly) is really nice for that!), everything LGTM on my end!

Wauplin commented:
Perfect! Thanks for the final pass, I'm finally merging this PR :smile:

julien-c commented on docs/source/guides/upload.md:
```diff
@@ -371,10 +377,91 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+      For example, json files can be merged into a single jsonl file, or large datasets can be exported as Parquet files.
+    - The maximum number of files per folder cannot exceed 10k files per folder. A simple solution is to
+      create a repository structure that uses subdirectories. For example, a repo with 1k folders from `000/` to `999/`, each containing at most 1000 files, is already enough.
+- **File size**: In the case of uploading large files (e.g. model weights), we strongly recommend splitting them **into chunks of around 5GB each**.
```
late to the party but i would have done larger, e.g. 20GB or at least 10GB (Cloudfront caches up to 30GB if i'm not mistaken)

@Pierrci @huggingface/moon-landing-back

Pierrci replied:
yes, but I don't really see the interest; as mentioned just below in the doc splitting into small chunks is better for uploading/downloading and retries, while I'm not sure there are a lot of advantages in doing chunks of 30GB?

severo commented:
@julien-c about https://discuss.huggingface.co/t/is-there-a-size-limit-for-dataset-hosting/14861/13?u=severo, we had the information in a previous version of this PR (see the table above: https://github.com/huggingface/huggingface_hub/pull/1565#issuecomment-1645548504).

Should we add it again?

julien-c commented:
i think we didn't want to do a table with "recommended limits" and "hard limits", so maybe just add a sentence like:

> in all cases no single LFS file will be able to be >50GB. I.e. 50GB is the hard limit for single file size.

severo commented:
OK: https://github.com/huggingface/huggingface_hub/pull/1624
